### PR TITLE
view can replace regions's el

### PR DESCRIPTION
**Pull request by ahumphreys87**
Thanks to @trezy for the original PR.

Ive added some tests and fixed an issue where `replaceView` could clash with `preventDestroy`

ping @marionettejs/marionette-core thoughts on getting this into 2.5 ready for CompositeView removal in 3.0?

### CODE REVIEW

**Comment by jasonLaster** [comment]
:nut_and_bolt: :wrench: :hammer:

**Comment by jmeas** [comment]
Curious to hear @jridgewell's thoughts on this implementation.

**Comment by ahumphreys87** [comment]
yup it's a worryingly simple implementation so would be good to get more opinions. I think the tests cover most cases though

File: src/region.js [hunk 1]
```diff
@@ -204,12 +208,38 @@ Marionette.Region = Marionette.Object.extend({
     return Backbone.$(el, Marionette._getValue(this.options.parentEl, this));
   },
 
+  // Replace the region's DOM element with the view's DOM element.
+  replaceEl: function(view) {
+    // empty el so we don't save any non-destroyed views
+    this.$el.contents().detach();
```
Comment (samccone) [inline src/region.js]: worrying about leaks here

File: test/unit/region.spec.js [hunk 2]
```diff
@@ -466,6 +497,65 @@ describe('region', function() {
...
+        });
+
+        it('should restore the region\'s "el" to the DOM', function() {
+          expect(this.$parentEl).to.contain.$html('<div id="region"></div>');
+        });
+      });
+
+      describe('and showing another view', function() {
+        beforeEach(function() {
+          this.MyView2 = Marionette.View.extend({
```
Comment (samccone) [inline test/unit/region.spec.js]: any reason why this is just a generic view?

File: src/region.js (hunk 1 above)
Comment (jridgewell) [inline src/region.js]: I'm curious why this is needed. The comment seems to point to a subview inside this regions element being `#remove`d, but that'll just work naturally with `$#remove`.

Comment (trezy) [inline src/region.js]: I'm with @samccone and @jridgewell. If you look at [my original PR](https://github.com/marionettejs/backbone.marionette/commit/fe0d47fdc6162484bff2c97918ef2720e05ea4a0#diff-b8da5292cb07ec757ac64e78ff5af5c1R143) I didn't do anything fancy. I just replaced the old `el` with the new `el` so that we retain references to everything but we're not storing anything unnecessary.

File: src/region.js [hunk 3]
```diff
@@ -204,12 +208,38 @@ Marionette.Region = Marionette.Object.extend({
...
   // appended to the `$el` that the region is managing
-  attachHtml: function(view) {
-    this.$el.contents().detach();
+  attachHtml: function(view, shouldReplace) {
+    if (shouldReplace) {
+      // replace the region's node with the view's node
+      this.replaceEl(view);
+    } else {
+      // empty the node and append new view
+      this.$el.contents().detach();
```
Comment (trezy) [inline src/region.js]: Same as @samccone's comment above. We should be rebinding events and such when we do this so `detach` could leave unwanted data in the nether whereas `remove` should cover our GC needs.

File: src/region.js [hunk 4]
```diff
@@ -204,12 +208,38 @@ Marionette.Region = Marionette.Object.extend({
...
+  },
+
   // Override this method to change how the new view is
   // appended to the `$el` that the region is managing
-  attachHtml: function(view) {
-    this.$el.contents().detach();
+  attachHtml: function(view, shouldReplace) {
+    if (shouldReplace) {
+      // replace the region's node with the view's node
+      this.replaceEl(view);
```
Comment (jridgewell) [inline src/region.js]: I'm not sure this is can be called twice, which is a valid usecase. Seems `#replaceEl` will try replacing the `this.el` with the new view's `el` and, since `this.el.parentNode = null` the second time around, an error will be thrown.

File: src/region.js (hunk 1 above)
Comment (ahumphreys87) [inline src/region.js]: Ok so this was added to handle the case where replaceEl was used with preventDestroy.

If we don't remove the old view in here when we restore it the old view will come back as we haven't destroyed it.

I just took this from attachHtml - I'll try with remove

File: src/region.js (hunk 3 above)
Comment (ahumphreys87) [inline src/region.js]: As above - detach is used in the current implementation of attachHtml I've just moved it to the else case here

File: test/unit/region.spec.js (hunk 2 above)
Comment (paulfalgout) [inline test/unit/region.spec.js]: This is `next` so `View === ItemView === LayoutView`?  and the old `View` is now `AbstractView`

**Comment by trezy** [comment]
I'm glad to finally see this getting into core! I won't have to use my shim anymore. :-)

I can also say that I've been using my shim since I made the original PR which would have been long enough for a child to gestate and I've had zero issues with it. Performance is good and I haven't seen any view collisions.

**Comment by jdaudier** [comment]
@trezy length of a child gestation... lol

File: src/region.js (hunk 1 above)
Comment (trezy) [inline src/region.js]: Alternatively we could detach and place the results in storage, we just need to make sure that view gets destroyed at some point for proper garbage collection.

Comment (ahumphreys87) [inline src/region.js]: yeah thats the thing with `preventDestroy` if that is used Mn has no power over memory leaks, personally I hate the option (views should be recreated) but it is used quite a bit.

I think for 2.5 we'll have to stick with `detach` here. For 3.0 im open to removing the `preventDestroy` function but that needs some discussion. Ill open an issue

File: src/region.js (hunk 4 above)
Comment (ahumphreys87) [inline src/region.js]: ah good catch - ill add a test case and fix!

File: src/region.js (hunk 1 above)
Comment (trezy) [inline src/region.js]: A better option would probably be to prevent using both options together. `replaceEl` + `preventDestroy` could throw an error or, at least, should throw a `console.info` that warns the dev about it. I hate the idea of building in a feature that almost encourages poor performance.

Comment (trezy) [inline src/region.js]: After talking to @ahumphreys87 in Gitter I think we're fine here. `preventDestroy` is a tool specifically for creating memory leaks so I think it's all good. ;-)

**Comment by ahumphreys87** [comment]
@samccone @jasonLaster bring it on!!!!!

File: src/region.js (hunk 4 above)
Comment (ahumphreys87) [inline src/region.js]: @jridgewell added test and modifed `replaceEl` to cater for this case.

**Comment by jasonLaster** [comment]
Looks good @ahumphreys87. :+1: Lets move forward into the future

**Comment by ahumphreys87** [comment]
@marionettejs/marionette-core needs 1 more :smile:

**Comment by paulfalgout** [comment]
I think `restoreEl` and `replaceEl` should probably be private.  Otherwise :+1:

**Comment by ahumphreys87** [comment]
@paulfalgout methods now private

**Comment by jasonLaster** [comment]
At long last :runner: :athletic_shoe:

**Comment by trezy** [comment]
Woooooooooohooooooooo! No more shims for meeeeee!

**Comment by howardroark** [comment]
Now I'm just waiting for the PR named "template can replace view's el" :) :)
